decidim-core/app/services/decidim/action_logger.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+module Decidim
+  # Use this class to log actions by any user.
+  #
+  # Usage:
+  #
+  #    ActionLogger.log(:create, user, new_proposal, extra_data)
+  class ActionLogger
+    # Public: Logs the given `action` by the given `user` on the given `resource`.
+    # Delegates the work to the instance method.
+    #
+    # action - a String representing the name of the action
+    # user - the Decidim::User that performed the action
+    # resource - the resource onn which the action was performed
+    # resource_extra - a Hash with resource_extra info to be recorded
+    #
+    # Returns the newly created `Decidim::ActionLog` resource.
+    def self.log(action, user, resource, resource_extra = {})
+      new(action, user, resource, resource_extra).log!
+    end
+
+    # Public: Initializes the instance.
+    #
+    # action - a String representing the name of the action
+    # user - the Decidim::User that performed the action
+    # resource - the resource onn which the action was performed
+    # resource_extra - a Hash with resource_extra info to be recorded
+    def initialize(action, user, resource, resource_extra = {})
+      @action = action
+      @user = user
+      @resource = resource
+      @resource_extra = resource_extra
+    end
+
+    # Public: Logs the given `action` by the given `user` on the given
+    # `resource`.
+    #
+    # Returns the newly created `Decidim::ActionLog` resource.
+    def log!
+      Decidim::ActionLog.create!(
+        user: user,
+        organization: organization,
+        action: action,
+        resource: resource,
+        participatory_space: participatory_space,
+        feature: feature,
+        extra: extra_data
+      )
+    end
+
+    private
+
+    attr_reader :action, :user, :resource, :resource_extra
+
+    def organization
+      user.organization
+    end
+
+    def feature
+      resource.feature if resource.respond_to?(:feature)
+    end
+
+    def participatory_space
+      resource.participatory_space if resource.respond_to?(:participatory_space)
```

Comment:
Do all resource delegate the participatory space to the feature?

If not, maybe you can try to get it from the feature too